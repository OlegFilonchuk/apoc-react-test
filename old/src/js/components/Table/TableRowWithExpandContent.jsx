import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import TableCellWithTextEllipsis from './TableCellWithTextEllipsis';
import ExpandButton from './ExpandButton';

import './TableRowWithExpandContent.less';

const emptyFn = () => null;

const clickableTags = ['BUTTON', 'A'];

function isClickableElementClicked(target, stopAtTagName) {
  let currentTarget = target;

  while (currentTarget.tagName !== stopAtTagName) {
    if (clickableTags.includes(currentTarget.tagName) && !currentTarget.classList.contains('expand-btn')) {
      return true;
    }

    currentTarget = currentTarget.parentNode;
  }

  return false;
}

export default class TableRowWithExpandContent extends React.Component {
  onExpand = e => {
    const target = e.nativeEvent.target;

    if (isClickableElementClicked(target, 'TR')) {
      return;
    }

    this.props.onExpand(this.props.id);
  };

  get expandBtn() {
    return (
      <ExpandButton onExpand={this.props.expandOnClick ? emptyFn : this.onExpand} isExpanded={this.props.isExpanded} />
    );
  }

  getCellForExpandedRow = cell => {
    const { children, className, ...rest } = cell.props;

    return this.props.isExpanded && cell.type === TableCellWithTextEllipsis ? (
      <td className={classNames('table-cell-with-expand-btn', className)} {...rest}>
        {this.expandBtn}
        {children}
      </td>
    ) : (
      React.cloneElement(cell, {
        className: classNames('table-cell-with-expand-btn', cell.props.className),
        children: React.Children.map(cell.props.children, element => [this.expandBtn, element])
      })
    );
  };

  get finalClassName() {
    return classNames(this.props.className, {
      'table-row-with-expand-content__clickable': this.props.expandOnClick
    });
  }

  getFinalContent = ({ columnWithButton, content, additionalProps = {} }) => (
    <tr
      {...additionalProps}
      className={this.finalClassName}
      onClickCapture={this.props.expandOnClick ? this.onExpand : emptyFn}
    >
      {columnWithButton}
      {content}
    </tr>
  );

  render() {
    const [firstColumn, ...restColumns] = React.Children.toArray(this.props.children);
    const columnWithButton = this.getCellForExpandedRow(firstColumn);
    const columns = [...restColumns];
    const content = React.Children.map(React.Children.toArray(columns), child =>
      this.props.isExpanded && child.type === TableCellWithTextEllipsis ? <td {...child.props} /> : child
    );

    return this.props.customWrapper
      ? this.props.customWrapper(props =>
          this.getFinalContent({
            columnWithButton,
            content,
            additionalProps: {
              ...props,
              ref: props.innerRef
            }
          })
        )
      : this.getFinalContent({ columnWithButton, content });
  }
}

TableRowWithExpandContent.propTypes = {
  /* eslint-disable react/no-unused-prop-types */
  /**
   * TableRowWithExpandContent's id
   */
  id: PropTypes.string.isRequired,

  /**
   * It will be called if expand button is clicked
   */
  onExpand: PropTypes.func,
  /* eslint-enable react/no-unused-prop-types */

  /**
   * List of custom classes added to component
   */
  className: PropTypes.string,

  /**
   * Array of table's cells
   */
  children: PropTypes.node.isRequired,

  /**
   * Show if row was expanded - used only by TableBodyWithExpandRows component internally
   */
  isExpanded: PropTypes.bool,

  /**
   * if true it triggers onExpand prop - used only by TableBodyWithExpandRows component internally
   */
  expandOnClick: PropTypes.bool,

  /**
   * Custom wrapper for table children. By default, table component and dependant expects particular children types
   * as they are operating on them. This props allow to add custom wrapper around them AFTER they are modified.
   */
  customWrapper: PropTypes.func
};

TableRowWithExpandContent.defaultProps = {
  className: '',
  onClick: null,
  isExpanded: false,
  expandOnClick: false,
  onExpand: () => null,
  customWrapper: null
};
