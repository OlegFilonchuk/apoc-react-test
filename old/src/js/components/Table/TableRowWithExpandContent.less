@import './tableVariables.less';
@import '~chameleon/less/variables';

@paddingLeftForCellWithExpandButton: calc(@expandButtonWidth + @cellPadding);

.expandable-table-row-trigger {
  &.expandable-table-row-trigger--expanded {
    td {
      background-color: @emptyTableContainerBackground;
    }
  }

  .table-cell-with-expand-btn {
    position: relative;
    padding-left: @paddingLeftForCellWithExpandButton;

    .expand-content-wrap {
      padding-left: @paddingLeftForCellWithExpandButton;
    }
  }

  .expand-btn {
    background: none;
    border: none;
    border-radius: 0;
    border-right: 1px solid @tableBorderColor;
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    width: @expandButtonWidth;
    z-index: 2;
  }
}

.table-row-with-expand-content__clickable {
  cursor: pointer;

  &:hover > td {
    background-color: @table-bg-hover;
  }
}
