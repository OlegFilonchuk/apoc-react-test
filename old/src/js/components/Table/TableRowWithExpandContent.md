Purpose of this component is to be used together with [TableBodyWithExpandRows](#tablehbodywithexpandrows). See example there.
