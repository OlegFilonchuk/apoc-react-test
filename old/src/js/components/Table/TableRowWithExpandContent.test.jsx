import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import TableRowWithExpandContent from './TableRowWithExpandContent';

const mountRow = ({ mountMethod = mount, props = {} }) =>
  mountMethod(
    <TableRowWithExpandContent {...props}>
      <td>Some other content</td>
      <td>
        <button id="clickable-content">Content</button>
        <a id="clickable-content2">Content2</a>
      </td>
      <td>
        <span id="not-clickable-content">Content</span>
      </td>
    </TableRowWithExpandContent>,
    {
      attachTo: document.createElement('tbody')
    }
  );

describe('<TableRowWithExpandContent />', () => {
  it('should match snapshot', () => {
    expect(toJson(mountRow({ mountMethod: shallow, props: { id: 'foo', isExpanded: false } }))).toMatchSnapshot();
  });

  it('correctly set `className`', () => {
    const cell = mountRow({ props: { className: 'some-extra-class', id: 'foo', isExpanded: false } });
    const element = cell.find('tr');

    expect(element).toHaveClassName('some-extra-class');
  });

  describe('expandOnClick', () => {
    let cell;
    const onExpandCallback = jest.fn();

    beforeAll(() => {
      cell = mountRow({
        props: {
          className: 'some-extra-class',
          id: 'foo',
          isExpanded: false,
          expandOnClick: true,
          onExpand: onExpandCallback
        }
      });
    });

    beforeEach(() => onExpandCallback.mockReset());

    describe('clicked on non-clickable element', () => {
      it('expands a row', () => {
        cell.find('#not-clickable-content').simulate('click');
        expect(onExpandCallback).toHaveBeenCalled();
      });
    });

    describe('clicked button', () => {
      it('does not expands a row', () => {
        cell.find('button#clickable-content').simulate('click');
        expect(onExpandCallback).not.toHaveBeenCalled();
      });
    });

    describe('clicked internal "expand button"', () => {
      it('does not expands a row', () => {
        cell.find('button.expand-btn').simulate('click');
        expect(onExpandCallback).toHaveBeenCalled();
      });
    });

    describe('clicked link', () => {
      it('does not expands a row', () => {
        cell.find('a#clickable-content2').simulate('click');
        expect(onExpandCallback).not.toHaveBeenCalled();
      });
    });
  });
});
