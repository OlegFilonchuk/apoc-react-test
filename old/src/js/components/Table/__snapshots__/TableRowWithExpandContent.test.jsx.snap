// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableRowWithExpandContent /> should match snapshot 1`] = `
<tr
  className=""
  onClickCapture={[Function]}
>
  <td
    className="table-cell-with-expand-btn"
    key=".0"
  >
    <ExpandButton
      className=""
      isExpanded={false}
      key=".0/.0"
      onExpand={[Function]}
      type="button"
    />
    Some other content
  </td>
  <td
    key=".$.$.1"
  >
    <button
      id="clickable-content"
    >
      Content
    </button>
    <a
      id="clickable-content2"
    >
      Content2
    </a>
  </td>
  <td
    key=".$.$.2"
  >
    <span
      id="not-clickable-content"
    >
      Content
    </span>
  </td>
</tr>
`;
